File: backend/routes/tasks.py
"""Task CRUD routes with pipeline execution controls."""

from __future__ import annotations

import asyncio
from typing import Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

import storage
from orchestrator import engine
from orchestrator.pipelines import list_pipelines, get_pipeline
from routes.websocket import broadcast

router = APIRouter()


class CreateTaskRequest(BaseModel):
    name: str
    description: Optional[str] = ""
    priority: Optional[str] = "medium"
    pipeline_id: Optional[str] = None


class UpdateTaskRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None


def _task_path(project_id: str, task_id: str):
    return storage.project_tasks_dir(project_id) / task_id / "task.json"


def _get_task(project_id: str, task_id: str):
    task = storage.read_json(_task_path(project_id, task_id))
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.post("/projects/{project_id}/tasks")
async def create_task(project_id: str, req: CreateTaskRequest):
    project = storage.read_json(storage.project_json_path(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    base_slug = storage.slugify(req.name)
    if not base_slug:
        raise HTTPException(status_code=400, detail="Task name produces an empty slug")
    task_id = storage.unique_task_slug(project_id, base_slug)

    task = {
        "id": task_id,
        "project_id": project_id,
        "name": req.name,
        "title": req.name,
        "description": req.description,
        "priority": req.priority,
        "status": "pending",
        "pipeline_id": req.pipeline_id,
        "branch_name": f"task/{task_id}",
        "current_agent": None,
        "current_step": None,
        "paused": False,
        "created_at": storage.now_iso(),
        "updated_at": storage.now_iso(),
    }

    task_dir = storage.project_tasks_dir(project_id) / task_id
    task_dir.mkdir(parents=True, exist_ok=True)
    storage.write_json(_task_path(project_id, task_id), task)

    return task


@router.get("/projects/{project_id}/tasks")
async def list_tasks(project_id: str):
    tasks_dir = storage.project_tasks_dir(project_id)
    if not tasks_dir.exists():
        return []

    tasks = []
    for task_dir in sorted(tasks_dir.iterdir()):
        if task_dir.is_dir():
            task_file = task_dir / "task.json"
            task = storage.read_json(task_file)
            if task:
                tasks.append(task)

    return sorted(tasks, key=lambda t: t.get("created_at", ""), reverse=True)


@router.get("/projects/{project_id}/tasks/{task_id}")
async def get_task(project_id: str, task_id: str):
    return _get_task(project_id, task_id)


@router.put("/projects/{project_id}/tasks/{task_id}")
async def update_task(project_id: str, task_id: str, req: UpdateTaskRequest):
    task = _get_task(project_id, task_id)
    updates = req.model_dump(exclude_none=True)
    task.update(updates)
    task["updated_at"] = storage.now_iso()
    storage.write_json(_task_path(project_id, task_id), task)
    return task


@router.delete("/projects/{project_id}/tasks/{task_id}")
async def delete_task(project_id: str, task_id: str):
    task_dir = storage.project_tasks_dir(project_id) / task_id
    if not task_dir.exists():
        raise HTTPException(status_code=404, detail="Task not found")

    # Clean up git worktree if one was created for this task
    task = storage.read_json(task_dir / "task.json")
    if task and task.get("worktree_path"):
        import worktree as wt
        repo_path = storage.get_repo_path(project_id)
        if repo_path:
            await wt.remove_worktree(repo_path, task_id)

    storage.delete_path(task_dir)
    return {"deleted": True}


# --- Pipeline endpoints ---


@router.get("/projects/{project_id}/pipelines")
async def get_pipelines(project_id: str):
    return list_pipelines(project_id)


class RunPipelineRequest(BaseModel):
    pipeline_id: str


@router.post("/projects/{project_id}/tasks/{task_id}/run")
async def run_task_pipeline(project_id: str, task_id: str, req: RunPipelineRequest):
    """Start running a task through a pipeline."""
    task = _get_task(project_id, task_id)

    pipeline = get_pipeline(project_id, req.pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    async def on_event(event: dict):
        await broadcast(project_id, event)

    # Run in background
    asyncio.create_task(engine.start_pipeline(project_id, task_id, req.pipeline_id, on_event))

    return {"status": "started", "pipeline": pipeline["name"]}


@router.post("/projects/{project_id}/tasks/{task_id}/pause")
async def pause_task(project_id: str, task_id: str):
    if not engine.pause_task(task_id):
        raise HTTPException(status_code=404, detail="Task not running")
    await broadcast(project_id, {"type": "task_paused", "task_id": task_id})
    return {"status": "paused"}


@router.post("/projects/{project_id}/tasks/{task_id}/resume")
async def resume_task(project_id: str, task_id: str):
    if not engine.resume_task(task_id):
        # Task not in memory (e.g. after server restart) — just update status on disk
        task = _get_task(project_id, task_id)
        task["status"] = "choosing_agent"
        task["paused"] = False
        task["updated_at"] = storage.now_iso()
        storage.write_json(storage.project_tasks_dir(project_id) / task_id / "task.json", task)
    await broadcast(project_id, {"type": "task_resumed", "task_id": task_id})
    return {"status": "running"}


@router.post("/projects/{project_id}/tasks/{task_id}/cancel")
async def cancel_task(project_id: str, task_id: str):
    # Try to cancel in-memory execution first
    engine.cancel_task(task_id)
    # Always update status on disk (handles stale tasks after server restart)
    task = _get_task(project_id, task_id)
    task["status"] = "cancelled"
    task["paused"] = False
    task["updated_at"] = storage.now_iso()
    storage.write_json(_task_path(project_id, task_id), task)
    await broadcast(project_id, {"type": "task_cancelled", "task_id": task_id})
    return {"status": "cancelled"}



class NextAgentRequest(BaseModel):
    agent: Optional[str] = None  # None means "done / finish pipeline"


@router.post("/projects/{project_id}/tasks/{task_id}/next-agent")
async def set_next_agent(project_id: str, task_id: str, req: NextAgentRequest):
    """User chooses which agent runs next, or None to finish."""
    if engine.set_next_agent(task_id, req.agent):
        # Task was in memory — pipeline loop continues
        await broadcast(project_id, {
            "type": "next_agent_chosen",
            "task_id": task_id,
            "agent": req.agent,
        })
        return {"status": "ok", "next_agent": req.agent}

    # Task not in memory (e.g. after server restart) — handle gracefully
    if req.agent is None:
        # User chose "Finish" — just mark as completed
        task = _get_task(project_id, task_id)
        task["status"] = "completed"
        task["paused"] = False
        task["updated_at"] = storage.now_iso()
        storage.write_json(storage.project_tasks_dir(project_id) / task_id / "task.json", task)
        await broadcast(project_id, {"type": "task_completed", "task_id": task_id})
        return {"status": "ok", "next_agent": None}

    # User chose an agent — start a fresh single-agent run
    async def on_event(event):
        await broadcast(project_id, event)

    asyncio.create_task(engine.run_single_agent(project_id, task_id, req.agent, on_event=on_event))
    await broadcast(project_id, {
        "type": "next_agent_chosen",
        "task_id": task_id,
        "agent": req.agent,
    })
    return {"status": "ok", "next_agent": req.agent}


class RunAgentRequest(BaseModel):
    agent: str
    context: Optional[str] = None


@router.post("/projects/{project_id}/tasks/{task_id}/run-agent")
async def run_agent(project_id: str, task_id: str, req: RunAgentRequest):
    """Run a specific agent on a task (works even on completed/cancelled tasks)."""
    _get_task(project_id, task_id)

    async def on_event(event: dict):
        await broadcast(project_id, event)

    asyncio.create_task(engine.run_single_agent(
        project_id, task_id, req.agent, req.context, on_event
    ))

    return {"status": "started", "agent": req.agent}


class AskAgentRequest(BaseModel):
    agent: str
    message: str


@router.post("/projects/{project_id}/tasks/{task_id}/ask-agent")
async def ask_agent(project_id: str, task_id: str, req: AskAgentRequest):
    """Ask an agent a question with full task context, without affecting pipeline state."""
    _get_task(project_id, task_id)

    async def on_event(event: dict):
        await broadcast(project_id, event)

    # Run in background so we return immediately; chunks stream via WebSocket
    asyncio.create_task(engine.ask_agent(project_id, task_id, req.agent, req.message, on_event))

    return {"status": "asking", "agent": req.agent}


class PermissionResponseRequest(BaseModel):
    permission_id: str
    behavior: str  # "allow" or "deny"
    message: Optional[str] = None


@router.post("/projects/{project_id}/tasks/{task_id}/permission-response")
async def permission_response(project_id: str, task_id: str, req: PermissionResponseRequest):
    """Respond to a permission request from an agent."""
    if not engine.resolve_permission(task_id, req.permission_id, req.behavior, req.message):
        raise HTTPException(status_code=404, detail="No pending permission with that ID")
    await broadcast(project_id, {
        "type": "permission_resolved",
        "task_id": task_id,
        "permission_id": req.permission_id,
        "behavior": req.behavior,
    })
    return {"status": "resolved", "behavior": req.behavior}


class InjectRequest(BaseModel):
    context: str


@router.post("/projects/{project_id}/tasks/{task_id}/inject")
async def inject_context(project_id: str, task_id: str, req: InjectRequest):
    """Inject extra instructions before the next step runs."""
    if not engine.inject_context(task_id, req.context):
        raise HTTPException(status_code=404, detail="Task not running")
    return {"status": "injected"}


@router.get("/projects/{project_id}/tasks/{task_id}/status")
async def task_execution_status(project_id: str, task_id: str):
    """Get live execution status of a running task."""
    status = engine.get_execution_status(task_id)
    if not status:
        task = _get_task(project_id, task_id)
        return {"task_id": task_id, "status": task.get("status", "unknown"), "running": False}
    return {**status, "running": True}


@router.get("/projects/{project_id}/tasks/{task_id}/history")
async def task_history(project_id: str, task_id: str):
    from orchestrator.models import get_history
    return get_history(project_id, task_id)


@router.get("/projects/{project_id}/tasks/{task_id}/artifacts")
async def task_artifacts(project_id: str, task_id: str):
    from orchestrator.models import list_artifacts
    return list_artifacts(project_id, task_id)


@router.get("/projects/{project_id}/tasks/{task_id}/artifacts/{artifact_type}/content")
async def get_artifact_content_route(
    project_id: str,
    task_id: str,
    artifact_type: str,
    run: Optional[str] = Query(default=None, description="Run selector: integer, 'latest', or omit for all"),
):
    """Get artifact content with optional run filter.

    ?run=latest  — latest run only
    ?run=1       — specific run number
    (omit)       — all runs (raw content)
    """
    from orchestrator.models import get_artifact_content

    run_param = None
    if run is not None:
        if run == "latest":
            run_param = "latest"
        else:
            try:
                run_param = int(run)
            except ValueError:
                raise HTTPException(status_code=400, detail="Invalid run parameter")

    content = get_artifact_content(project_id, task_id, artifact_type, run=run_param)
    if content is None:
        raise HTTPException(status_code=404, detail="Artifact not found")
    return {"artifact_type": artifact_type, "run": run, "content": content}


@router.get("/projects/{project_id}/tasks/{task_id}/artifacts/{artifact_type}/runs")
async def get_artifact_runs(project_id: str, task_id: str, artifact_type: str):
    """Return run metadata list parsed from artifact file delimiters."""
    from orchestrator.models import get_artifact_content, split_runs
    import storage as _storage

    artifacts_dir = _storage.project_tasks_dir(project_id) / task_id / "artifacts"
    content_path = artifacts_dir / f"{artifact_type}.md"
    if not content_path.exists():
        raise HTTPException(status_code=404, detail="Artifact not found")

    raw = _storage.read_text_file(content_path)
    runs = split_runs(raw)

    run_meta = [
        {"run": r["run"], "agent": r["agent"], "timestamp": r["timestamp"]}
        for r in runs
    ]
    return {
        "artifact_type": artifact_type,
        "run_count": len(runs),
        "runs": run_meta,
    }


class UpdateArtifactRequest(BaseModel):
    content: str


@router.put("/projects/{project_id}/tasks/{task_id}/artifacts/{artifact_type}")
async def update_artifact(
    project_id: str, task_id: str, artifact_type: str, req: UpdateArtifactRequest
):
    """Edit an artifact's content (human-in-the-loop editing before passing downstream)."""
    from orchestrator.models import update_artifact_content
    result = update_artifact_content(project_id, task_id, artifact_type, req.content)
    if not result:
        raise HTTPException(status_code=404, detail="Artifact not found")
    return result


@router.get("/projects/{project_id}/tasks/{task_id}/terminals")
async def task_terminals(project_id: str, task_id: str):
    """Get saved terminal messages for a task."""
    from orchestrator.models import load_terminals
    return load_terminals(project_id, task_id)

File: backend/src/routes/tasks.ts
/**
 * Task CRUD routes with pipeline execution controls.
 */

import path from "path";
import * as storage from "../storage.js";
import * as engine from "../orchestrator/engine.js";
import * as wt from "../worktree.js";
import { listPipelines, getPipeline } from "../orchestrator/pipelines.js";
import { listArtifacts, updateArtifactContent, getHistory, loadTerminals } from "../orchestrator/models.js";
import { AGENT_NAMES } from "../agents/registry.js";
import type { FastifyInstance } from "fastify";

const VALID_AGENT_NAMES = AGENT_NAMES as readonly string[];

type Broadcast = (projectId: string, event: Record<string, unknown>) => Promise<void>;

function taskPath(projectId: string, taskId: string): string {
  return path.join(storage.projectTasksDir(projectId), taskId, "task.json");
}

async function getTask(
  projectId: string,
  taskId: string,
): Promise<Record<string, unknown> | null> {
  return storage.readJson<Record<string, unknown>>(taskPath(projectId, taskId));
}

export async function registerTaskRoutes(
  app: FastifyInstance,
  broadcast: Broadcast,
): Promise<void> {
  // POST /projects/:projectId/tasks
  app.post<{
    Params: { projectId: string };
    Body: { name: string; description?: string; priority?: string; pipeline_id?: string };
  }>("/projects/:projectId/tasks", async (req, reply) => {
    const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
    if (!project) return reply.code(404).send({ detail: "Project not found" });

    const baseSlug = storage.slugify(req.body.name);
    if (!baseSlug) return reply.code(400).send({ detail: "Task name produces an empty slug" });
    const taskId = await storage.uniqueTaskSlug(req.params.projectId, baseSlug);

    const task = {
      id: taskId,
      project_id: req.params.projectId,
      name: req.body.name,
      title: req.body.name,
      description: req.body.description ?? "",
      priority: req.body.priority ?? "medium",
      status: "pending",
      pipeline_id: req.body.pipeline_id ?? null,
      branch_name: "task/" + taskId,
      current_agent: null,
      current_step: null,
      paused: false,
      created_at: storage.nowIso(),
      updated_at: storage.nowIso(),
    };

    const taskDir = path.join(storage.projectTasksDir(req.params.projectId), taskId);
    const { mkdir } = await import("fs/promises");
    await mkdir(taskDir, { recursive: true });
    await storage.writeJson(taskPath(req.params.projectId, taskId), task);

    reply.code(201);
    return task;
  });

  // GET /projects/:projectId/tasks
  app.get<{ Params: { projectId: string } }>("/projects/:projectId/tasks", async (req, reply) => {
    const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
    if (!project) return reply.code(404).send({ detail: "Project not found" });

    const { readdir } = await import("fs/promises");
    const tasksDir = storage.projectTasksDir(req.params.projectId);
    try {
      const entries = await readdir(tasksDir, { withFileTypes: true });
      const tasks: Record<string, unknown>[] = [];
      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        const task = await storage.readJson<Record<string, unknown>>(
          path.join(tasksDir, entry.name, "task.json"),
        );
        if (task) tasks.push(task);
      }
      return tasks.sort((a, b) =>
        String(b.created_at ?? "").localeCompare(String(a.created_at ?? "")),
      );
    } catch {
      return [];
    }
  });

  // GET /projects/:projectId/tasks/:taskId
  app.get<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId",
    async (req, reply) => {
      const task = await getTask(req.params.projectId, req.params.taskId);
      if (!task) return reply.code(404).send({ detail: "Task not found" });
      return task;
    },
  );

  // PUT /projects/:projectId/tasks/:taskId
  app.put<{
    Params: { projectId: string; taskId: string };
    Body: Record<string, unknown>;
  }>("/projects/:projectId/tasks/:taskId", async (req, reply) => {
    const task = await getTask(req.params.projectId, req.params.taskId);
    if (!task) return reply.code(404).send({ detail: "Task not found" });

    const allowed = ["title", "description", "priority", "status"];
    for (const key of allowed) {
      if (req.body[key] !== undefined) task[key] = req.body[key];
    }
    task.updated_at = storage.nowIso();
    await storage.writeJson(taskPath(req.params.projectId, req.params.taskId), task);
    return task;
  });

  // DELETE /projects/:projectId/tasks/:taskId
  app.delete<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId",
    async (req, reply) => {
      const taskDir = path.join(
        storage.projectTasksDir(req.params.projectId),
        req.params.taskId,
      );
      const { existsSync } = await import("fs");
      if (!existsSync(taskDir)) return reply.code(404).send({ detail: "Task not found" });

      // Clean up git worktree if one was created for this task
      const task = await storage.readJson<Record<string, unknown>>(
        path.join(taskDir, "task.json"),
      );
      if (task?.worktree_path) {
        const repoPath = storage.getRepoPath(req.params.projectId);
        if (repoPath) {
          await wt.removeWorktree(repoPath, req.params.taskId);
        }
      }

      await storage.deletePath(taskDir);
      return { deleted: true };
    },
  );

  // GET /projects/:projectId/pipelines
  app.get<{ Params: { projectId: string } }>("/projects/:projectId/pipelines", async (req) => {
    return listPipelines(req.params.projectId);
  });

  // POST /projects/:projectId/tasks/:taskId/run
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { pipeline_id: string };
  }>("/projects/:projectId/tasks/:taskId/run", async (req, reply) => {
    const task = await getTask(req.params.projectId, req.params.taskId);
    if (!task) return reply.code(404).send({ detail: "Task not found" });

    const pipeline = await getPipeline(req.params.projectId, req.body.pipeline_id);
    if (!pipeline) return reply.code(404).send({ detail: "Pipeline not found" });

    const { projectId, taskId } = req.params;
    const pipelineId = req.body.pipeline_id;

    // Run in background — don't await
    setImmediate(() => {
      engine
        .startPipeline(projectId, taskId, pipelineId, (event) => broadcast(projectId, event))
        .catch(console.error);
    });

    return { status: "started", pipeline: pipeline.name };
  });

  // POST /projects/:projectId/tasks/:taskId/pause
  app.post<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/pause",
    async (req, reply) => {
      const taskExists = await getTask(req.params.projectId, req.params.taskId);
      if (!taskExists) return reply.code(404).send({ detail: "Task not found" });
      if (!engine.pauseTask(req.params.taskId)) {
        return reply.code(400).send({ detail: "Task not running" });
      }
      await broadcast(req.params.projectId, {
        type: "task_paused",
        task_id: req.params.taskId,
      });
      return { status: "paused" };
    },
  );

  // POST /projects/:projectId/tasks/:taskId/resume
  app.post<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/resume",
    async (req, reply) => {
      const task = await getTask(req.params.projectId, req.params.taskId);
      if (!task) return reply.code(404).send({ detail: "Task not found" });
      if (!engine.resumeTask(req.params.taskId)) {
        // Task not in memory (e.g. after server restart) — update status on disk
        task.status = "choosing_agent";
        task.paused = false;
        task.updated_at = storage.nowIso();
        await storage.writeJson(taskPath(req.params.projectId, req.params.taskId), task);
      }
      await broadcast(req.params.projectId, {
        type: "task_resumed",
        task_id: req.params.taskId,
      });
      return { status: "running" };
    },
  );

  // POST /projects/:projectId/tasks/:taskId/cancel
  app.post<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/cancel",
    async (req, reply) => {
      const task = await getTask(req.params.projectId, req.params.taskId);
      if (!task) return reply.code(404).send({ detail: "Task not found" });

      const terminalStates = ["cancelled", "completed", "failed"];
      if (terminalStates.includes(String(task.status ?? ""))) {
        return reply.code(400).send({ detail: "Task is already in terminal state" });
      }

      engine.cancelTask(req.params.taskId);

      task.status = "cancelled";
      task.paused = false;
      task.updated_at = storage.nowIso();
      await storage.writeJson(taskPath(req.params.projectId, req.params.taskId), task);

      await broadcast(req.params.projectId, {
        type: "task_cancelled",
        task_id: req.params.taskId,
      });
      return { status: "cancelled" };
    },
  );

  // POST /projects/:projectId/tasks/:taskId/next-agent
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { agent?: string | null };
  }>("/projects/:projectId/tasks/:taskId/next-agent", async (req, reply) => {
    const task = await getTask(req.params.projectId, req.params.taskId);
    if (!task) return reply.code(404).send({ detail: "Task not found" });

    const { projectId, taskId } = req.params;
    const agent = req.body.agent ?? null;

    if (engine.setNextAgent(taskId, agent)) {
      // Task was in memory — pipeline loop continues
      await broadcast(projectId, { type: "next_agent_chosen", task_id: taskId, agent });
      return { status: "ok", next_agent: agent };
    }

    // Task not in memory (e.g. after server restart) — handle gracefully
    if (agent === null) {
      // User chose "Finish" — mark as completed
      task.status = "completed";
      task.paused = false;
      task.updated_at = storage.nowIso();
      await storage.writeJson(taskPath(projectId, taskId), task);
      await broadcast(projectId, { type: "task_completed", task_id: taskId });
      return { status: "ok", next_agent: null };
    }

    // User chose an agent — start a fresh single-agent run
    setImmediate(() => {
      engine
        .runSingleAgent(projectId, taskId, agent, null, (event) => broadcast(projectId, event))
        .catch(console.error);
    });
    await broadcast(projectId, { type: "next_agent_chosen", task_id: taskId, agent });
    return { status: "ok", next_agent: agent };
  });

  // POST /projects/:projectId/tasks/:taskId/run-agent
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { agent: string; context?: string };
  }>("/projects/:projectId/tasks/:taskId/run-agent", async (req, reply) => {
    const task = await getTask(req.params.projectId, req.params.taskId);
    if (!task) return reply.code(404).send({ detail: "Task not found" });

    const { projectId, taskId } = req.params;
    const { agent, context } = req.body;

    if (!VALID_AGENT_NAMES.includes(agent)) {
      return reply.code(400).send({
        detail: `Invalid agent name. Must be one of: ${VALID_AGENT_NAMES.join(", ")}`,
      });
    }

    setImmediate(() => {
      engine
        .runSingleAgent(projectId, taskId, agent, context ?? null, (event) =>
          broadcast(projectId, event),
        )
        .catch(console.error);
    });

    return { status: "started", agent };
  });

  // POST /projects/:projectId/tasks/:taskId/ask-agent
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { agent: string; message: string };
  }>("/projects/:projectId/tasks/:taskId/ask-agent", async (req, reply) => {
    const task = await getTask(req.params.projectId, req.params.taskId);
    if (!task) return reply.code(404).send({ detail: "Task not found" });

    const { projectId, taskId } = req.params;
    const { agent, message } = req.body;

    if (!VALID_AGENT_NAMES.includes(agent)) {
      return reply.code(400).send({
        detail: `Invalid agent name. Must be one of: ${VALID_AGENT_NAMES.join(", ")}`,
      });
    }

    setImmediate(() => {
      engine
        .askAgent(projectId, taskId, agent, message, (event) => broadcast(projectId, event))
        .catch(console.error);
    });

    return { status: "asking", agent };
  });

  // POST /projects/:projectId/tasks/:taskId/permission-response
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { permission_id: string; behavior: "allow" | "deny"; message?: string };
  }>("/projects/:projectId/tasks/:taskId/permission-response", async (req, reply) => {
    const { permission_id, behavior, message } = req.body;
    if (!engine.resolvePermission(req.params.taskId, permission_id, behavior, message)) {
      return reply.code(404).send({ detail: "No pending permission with that ID" });
    }
    await broadcast(req.params.projectId, {
      type: "permission_resolved",
      task_id: req.params.taskId,
      permission_id,
      behavior,
    });
    return { status: "resolved", behavior };
  });

  // POST /projects/:projectId/tasks/:taskId/inject
  app.post<{
    Params: { projectId: string; taskId: string };
    Body: { context: string };
  }>("/projects/:projectId/tasks/:taskId/inject", async (req, reply) => {
    const taskExists = await getTask(req.params.projectId, req.params.taskId);
    if (!taskExists) return reply.code(404).send({ detail: "Task not found" });
    if (!engine.injectContext(req.params.taskId, req.body.context)) {
      return reply.code(400).send({ detail: "Task not running" });
    }
    return { status: "injected" };
  });

  // GET /projects/:projectId/tasks/:taskId/status
  app.get<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/status",
    async (req, reply) => {
      const status = engine.getExecutionStatus(req.params.taskId);
      if (!status) {
        const task = await getTask(req.params.projectId, req.params.taskId);
        if (!task) return reply.code(404).send({ detail: "Task not found" });
        return { task_id: req.params.taskId, status: task.status ?? "unknown", running: false };
      }
      return { ...status, running: true };
    },
  );

  // GET /projects/:projectId/tasks/:taskId/history
  app.get<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/history",
    async (req) => {
      return getHistory(req.params.projectId, req.params.taskId);
    },
  );

  // GET /projects/:projectId/tasks/:taskId/artifacts
  app.get<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/artifacts",
    async (req) => {
      return listArtifacts(req.params.projectId, req.params.taskId);
    },
  );

  // PUT /projects/:projectId/tasks/:taskId/artifacts/:artifactType
  app.put<{
    Params: { projectId: string; taskId: string; artifactType: string };
    Body: { content: string };
  }>("/projects/:projectId/tasks/:taskId/artifacts/:artifactType", async (req, reply) => {
    const result = await updateArtifactContent(
      req.params.projectId,
      req.params.taskId,
      req.params.artifactType,
      req.body.content,
    );
    if (!result) return reply.code(404).send({ detail: "Artifact not found" });
    return result;
  });

  // GET /projects/:projectId/tasks/:taskId/terminals
  app.get<{ Params: { projectId: string; taskId: string } }>(
    "/projects/:projectId/tasks/:taskId/terminals",
    async (req) => {
      return loadTerminals(req.params.projectId, req.params.taskId);
    },
  );
}
